runtime/compiler/env/j9methodServer.cpp

```diff
@@ -128,23 +124,30 @@ TR_ResolvedJ9JITaaSServerMethod::getClassLoader()
 bool
 TR_ResolvedJ9JITaaSServerMethod::staticAttributes(TR::Compilation * comp, I_32 cpIndex, void * * address, TR::DataType * type, bool * volatileP, bool * isFinal, bool * isPrivate, bool isStore, bool * unresolvedInCP, bool needAOTValidation)
    {
-   // only make a query if it's not in the cache
-   auto gotAttrs = _staticAttributesCache.find(cpIndex); 
-   if (gotAttrs == _staticAttributesCache.end())
+   bool isStatic = true;
+   TR_J9MethodFieldAttributes attributes;
+   if(!getCachedFieldAttributes(cpIndex, &attributes, isStatic))
       {
+      // make a remote call, if attributes are not cached
       _stream->write(JITaaS::J9ServerMessageType::ResolvedMethod_staticAttributes, _remoteMirror, cpIndex, isStore, needAOTValidation);
       auto recv = _stream->read<TR_J9MethodFieldAttributes>();
-      gotAttrs = _staticAttributesCache.insert({cpIndex, std::get<0>(recv)}).first;
-      }
-   // access attributes from the cache
-   auto attrs = gotAttrs->second;
-   *address = attrs.resolvedFieldAttribute.address;
-   *type = attrs.resolvedFieldAttribute.type;
-   *volatileP = attrs.resolvedFieldAttribute.volatileP;
-   if (isFinal) *isFinal = attrs.resolvedFieldAttribute.isFinal;
-   if (isPrivate) *isPrivate = attrs.resolvedFieldAttribute.isPrivate;
-   if (unresolvedInCP) *unresolvedInCP = attrs.resolvedFieldAttribute.unresolvedInCP;
-   return attrs.resolvedFieldAttribute.result;
+      attributes = std::get<0>(recv);
+      cacheFieldAttributes(cpIndex, attributes, isStatic);
```

Comment:
I think there could be a race here. Two threads could check `getCachedFieldAttributes`, find that nothing is cached and then both execute (one after another) `cacheFieldAttributes`. Maybe if the information they write is the same it does not matter.
We could do another check in `cacheFieldAttributes` while we have the monitor, that indeed there is no information stored in there.

Reply:
I think information should be the same, otherwise we shouldn't even cache it.
I'll add the check though.